catalog/spatial/geocoding/spatial-geocoding-offline-catalog/src/main/java/org/codice/ddf/spatial/geocoding/index/GeoNamesCatalogIndexer.java

```diff
@@ -196,69 +219,110 @@ public void updateProgress(final int progress) {
+        List<Serializable> metacardsToDelete = null;
 
-      if (CollectionUtils.isEmpty(metacardsToDelete)) {
-        LOGGER.debug("No metacards to delete.");
-        return;
-      }
+        try {
+          QueryResponse response = catalogFramework.query(queryRequest);
+          metacardsToDelete =
+              response
+                  .getResults()
+                  .stream()
+                  .map(Result::getMetacard)
+                  .map(Metacard::getId)
+                  .collect(Collectors.toList());
+        } catch (Exception e) {
+          LOGGER.debug("Unable to query for metacards.", e);
+          errorOccurred = true;
+        }
 
-      for (int i = 0; i < metacardsToDelete.size(); i += BATCH_SIZE) {
+        if (!errorOccurred && CollectionUtils.isEmpty(metacardsToDelete)) {
+          break;
+        }
 
-        int lastIndex = i + BATCH_SIZE;
-        if (lastIndex > metacardsToDelete.size()) {
-          lastIndex = metacardsToDelete.size();
+        if (!errorOccurred) {
+          LOGGER.trace("Deleting {} GeoNames metacards", metacardsToDelete.size());
+
+          try {
+            removeMetacards(catalogProvider, extractionCallback, metacardsToDelete);
+          } catch (IngestException e) {
+            LOGGER.debug("Unable to delete metacards.", e);
+            errorOccurred = true;
+          }
         }
 
-        List<Serializable> sublist = metacardsToDelete.subList(i, lastIndex);
+        retries++;
 
-        DeleteRequest deleteRequest = new DeleteRequestImpl(sublist, Core.ID, new HashMap<>());
-        try {
-          DeleteResponse deleteResponse = catalogProvider.delete(deleteRequest);
-          List<Metacard> deletedMetacards = deleteResponse.getDeletedMetacards();
-          LOGGER.debug(
-              "{} metacards deleted.", deletedMetacards == null ? 0 : deletedMetacards.size());
-          extractionCallback.updateProgress(
-              (int) (((double) (i + 1) / metacardsToDelete.size()) * 50));
-        } catch (IngestException e) {
-          LOGGER.debug("Unable to delete metacards.", e);
+        if (errorOccurred && retries < maxRetries) {
+          try {
+            Thread.sleep(WAIT_TIME_MSEC);
+          } catch (InterruptedException ignore) {
+          }
         }
       }
     }
     extractionCallback.updateProgress(50);
   }
 
+  private void removeMetacards(
+      CatalogProvider catalogProvider,
+      ProgressCallback extractionCallback,
+      List<Serializable> metacards)
+      throws IngestException {
+    for (int i = 0; i < metacards.size(); i += BATCH_SIZE) {
+      int lastIndex = i + BATCH_SIZE;
+      if (lastIndex > metacards.size()) {
+        lastIndex = metacards.size();
+      }
+
+      List<Serializable> sublist = metacards.subList(i, lastIndex);
+
+      DeleteRequest deleteRequest = new DeleteRequestImpl(sublist, Core.ID, new HashMap<>());
+      DeleteResponse deleteResponse = catalogProvider.delete(deleteRequest);
```

Comment:
Should we try/catch here?  Is it possible that the response is null coming back?

Reply:
I think if the interface isn't marked it shouldn't return null in the catalog API